### PR TITLE
[SHOT] first pass at a builder

### PR DESCRIPTION
**Pull request by odow**
## Failures

- [x] `apple-darwin`
```
[05:31:30] /workspace/srcdir/SHOT/src/Model/../Model/Simplifications.h:1370:26: error: 'value' is unavailable: introduced in macOS 10.14
[05:31:30]                 optional.value()->coefficient *= -1.0;
[05:31:30]                          ^
[05:31:30] /opt/x86_64-apple-darwin14/x86_64-apple-darwin14/sys-root/usr/include/c++/v1/optional:947:27: note: 'value' has been explicitly marked unavailable here
[05:31:30]     constexpr value_type& value() &
[05:31:30]                           ^
```

- [x] `mingw`
```
[05:43:58] CMake Error at misc/FindFilesystem.cmake:61 (message):
[05:43:58]   No C++ support for std::filesystem
[05:43:58] Call Stack (most recent call first):
[05:43:58]   CMakeLists.txt:23 (find_package)
```

### CODE REVIEW

File: C/Coin-OR/SHOT/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    -DHAS_AMPL=on \
+    -DGENERATE_EXE=on \
+    ..
+
+make
+make install
+"""
+
+products = [
+    ExecutableProduct("SHOT", :amplexe),
```
Comment (giordano) [inline C/Coin-OR/SHOT/build_tarballs.jl:36]: Is the variable associated to the product, `amplexe`, correct?

**Comment by giordano** [comment]
For macOS, we provide SDK 10.12
https://github.com/JuliaPackaging/Yggdrasil/blob/4d4cfc1094287e1a239cdea674e96f0953a2994d/0_RootFS/PlatformSupport/build_tarballs.jl#L39
Does this mean they require v10.14?

File: C/Coin-OR/SHOT/build_tarballs.jl (hunk 1 above)
Comment (odow) [reply C/Coin-OR/SHOT/build_tarballs.jl:36]: Yeah, this is the convention we have gone with for Ipopt, Bonmin, and Couenne. It's a mess, but you use the SHOT executable by passing it an AMPL file.

**Comment by odow** [comment]
> Does this mean they require v10.14?

I guess so. Or maybe we just need a newer version of Clang? Can we update the compilers in the build script? Or are they locked with some magic?

**Comment by giordano** [comment]
The compiler is independent from the macOS SDK.  On macOS by default we use the latest version available, which at the moment is version 11 (latest release version is v12).  However, we only provide a single macOS SDK, a relatively old one to be compatible with all macOS versions supported by Julia, similar to the fact we ship an old glibc version on Linux, compatible with all supported systems.

**Comment by giordano** [comment]
For Windows, the error
```
[05:43:58] CMake Error at misc/FindFilesystem.cmake:61 (message):
[05:43:58]   No C++ support for std::filesystem
[05:43:58] Call Stack (most recent call first):
[05:43:58]   CMakeLists.txt:23 (find_package)
```
is due to a bug in GCC 8: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=78870, it can be fixed by requiring GCC 9.  BTW, can we require Julia v1.6 for this package?  It'd make requiring GCC 9 a bit safer

File: C/Coin-OR/SHOT/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    -DCMAKE_BUILD_TYPE=Release \
+    -DHAS_CBC=on \
+    -DCBC_DIR=${prefix} \
+    -DHAS_IPOPT=on \
+    -DIPOPT_DIR=${prefix} \
+    -DHAS_AMPL=on \
+    -DGENERATE_EXE=on \
+    ..
+
+make
```
Comment (giordano) [inline C/Coin-OR/SHOT/build_tarballs.jl:31]: Any particular reason why this isn't run in parallel? :sweat_smile:

Comment (odow) [reply C/Coin-OR/SHOT/build_tarballs.jl]: Nope. I'm just blind... and I wasn't looking at how long the builds were taking...

**Comment by giordano** [comment]
New error on Windows:
```
[23:16:45] /usr/bin/ccache /opt/bin/x86_64-w64-mingw32-libgfortran5-cxx11/x86_64-w64-mingw32-g++ --sysroot=/opt/x86_64-w64-mingw32/x86_64-w64-mingw32/sys-root/  -O3 -DNDEBUG  -shared -o libSHOTSolver.dll -Wl,--out-implib,libSHOTSolver.dll.a -Wl,--major-image-version,0,--minor-image-version,0 -Wl,--whole-archive CMakeFiles/SHOTSolver.dir/objects.a -Wl,--no-whole-archive @CMakeFiles/SHOTSolver.dir/linklibs.rsp
[23:16:48] /opt/x86_64-w64-mingw32/bin/../lib/gcc/x86_64-w64-mingw32/9.1.0/../../../../x86_64-w64-mingw32/bin/ld: cannot find -lasl
[23:16:48] collect2: error: ld returned 1 exit status
[23:16:48] make[2]: *** [CMakeFiles/SHOTSolver.dir/build.make:153: libSHOTSolver.dll] Error 1
[23:16:48] make[2]: Leaving directory '/workspace/srcdir/SHOT/build'
[23:16:48] make[1]: *** [CMakeFiles/Makefile2:451: CMakeFiles/SHOTSolver.dir/all] Error 2
```

**Comment by giordano** [comment]
New error on macOS:
```
[23:24:30] /usr/bin/ccache /opt/bin/x86_64-apple-darwin14-libgfortran5-cxx11/x86_64-apple-darwin14-clang++ --sysroot=/opt/x86_64-apple-darwin14/x86_64-apple-darwin14/sys-root -mmacosx-version-min=10.14 -O3 -DNDEBUG -dynamiclib -Wl,-headerpad_max_install_names  -o libSHOTSolver.dylib -install_name @rpath/libSHOTSolver.dylib CMakeFiles/SHOTSolver.dir/src/Report.cpp.o CMakeFiles/SHOTSolver.dir/src/RootsearchMethod/RootsearchMethodBoost.cpp.o CMakeFiles/SHOTSolver.dir/src/Solver.cpp.o  libSHOTHelper.a libSHOTModel.a libSHOTResults.a libSHOTDualStrategy.a libSHOTPrimalStrategy.a libSHOTTasks.a -Wl,--whole-archive libSHOTModelingInterfaces.a -Wl,--no-whole-archive libSHOTSolutionStrategies.a libSHOTTasks.a libSHOTDualStrategy.a -lCbcSolver -lCbc -lpthread -lasl -lCgl -lOsiClp -lClp -lOsi -lCoinUtils libSHOTPrimalStrategy.a libSHOTModel.a libSHOTHelper.a libtinyxml2.a -lipopt 
[23:24:30] ld: unknown option: --whole-archive
[23:24:30] clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
[23:24:30] make[2]: *** [CMakeFiles/SHOTSolver.dir/build.make:148: libSHOTSolver.dylib] Error 1
```
I deduce no one has ever built SHOT on macOS?

**Comment by giordano** [comment]
And after fixing the linker arguments for macOS I get:
```
[ 99%] Linking CXX shared library libSHOTSolver.dylib
/usr/bin/cmake -E cmake_link_script CMakeFiles/SHOTSolver.dir/link.txt --verbose=true
/usr/bin/ccache /opt/bin/x86_64-apple-darwin14-libgfortran5-cxx11/x86_64-apple-darwin14-clang++ --sysroot=/opt/x86_64-apple-darwin14/x86_64-apple-darwin14/sys-root -mmacosx-version-min=10.14 -O3 -DNDEBUG -dynamiclib -Wl,-headerpad_max_install_names  -o libSHOTSolver.dylib -install_name @rpath/libSHOTSolver.dylib CMakeFiles/SHOTSolver.dir/src/Report.cpp.o CMakeFiles/SHOTSolver.dir/src/RootsearchMethod/RootsearchMethodBoost.cpp.o CMakeFiles/SHOTSolver.dir/src/Solver.cpp.o  libSHOTHelper.a libSHOTModel.a libSHOTResults.a libSHOTDualStrategy.a libSHOTPrimalStrategy.a libSHOTTasks.a -Wl,-all_load libSHOTModelingInterfaces.a libSHOTSolutionStrategies.a libSHOTTasks.a libSHOTDualStrategy.a -lCbcSolver -lCbc -lpthread -lasl -lCgl -lOsiClp -lClp -lOsi -lCoinUtils libSHOTPrimalStrategy.a libSHOTModel.a libSHOTHelper.a libtinyxml2.a -lipopt 
ccache /opt/x86_64-linux-musl/bin/clang++ -target x86_64-apple-darwin14 --sysroot=/opt/x86_64-apple-darwin14/x86_64-apple-darwin14/sys-root -march=x86-64 -mtune=generic -nostdinc++ -isystem /opt/x86_64-apple-darwin14/x86_64-apple-darwin14/sys-root/usr/include/c++/v1 -Wno-unused-command-line-argument -mmacosx-version-min=10.10 --sysroot=/opt/x86_64-apple-darwin14/x86_64-apple-darwin14/sys-root -mmacosx-version-min=10.14 -O3 -DNDEBUG -dynamiclib -Wl,-headerpad_max_install_names -o libSHOTSolver.dylib -install_name @rpath/libSHOTSolver.dylib CMakeFiles/SHOTSolver.dir/src/Report.cpp.o CMakeFiles/SHOTSolver.dir/src/RootsearchMethod/RootsearchMethodBoost.cpp.o CMakeFiles/SHOTSolver.dir/src/Solver.cpp.o libSHOTHelper.a libSHOTModel.a libSHOTResults.a libSHOTDualStrategy.a libSHOTPrimalStrategy.a libSHOTTasks.a -Wl,-all_load libSHOTModelingInterfaces.a libSHOTSolutionStrategies.a libSHOTTasks.a libSHOTDualStrategy.a -lCbcSolver -lCbc -lpthread -lasl -lCgl -lOsiClp -lClp -lOsi -lCoinUtils libSHOTPrimalStrategy.a libSHOTModel.a libSHOTHelper.a libtinyxml2.a -lipopt -L/opt/x86_64-apple-darwin14/x86_64-apple-darwin14/lib -fuse-ld=x86_64-apple-darwin14 -headerpad_max_install_names
Undefined symbols for architecture x86_64:
  "__ZNKSt3__14__fs10filesystem4path10__filenameEv", referenced from:
      __ZN4SHOT6Solver10setProblemENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in Solver.cpp.o
      __ZN4SHOT6Solver19initializeDebugModeEv in Solver.cpp.o
      __ZN4SHOT6Solver10setProblemENSt3__110shared_ptrINS_7ProblemEEES4_NS2_INS_15IModelingSystemEEE in Solver.cpp.o
      __ZN4SHOT6Solver12solveProblemEv in Solver.cpp.o
      __ZNSt3__14__fs10filesystemdvERKNS1_4pathES4_ in libSHOTHelper.a(Utilities.cpp.o)
      __ZN4SHOT13NLPSolverSHOT20initializeMIPProblemEv in libSHOTPrimalStrategy.a(NLPSolverSHOT.cpp.o)
  "__ZNKSt3__14__fs10filesystem4path11__extensionEv", referenced from:
      __ZN4SHOT6Solver18setOptionsFromFileENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in Solver.cpp.o
      __ZN4SHOT6Solver10setProblemENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in Solver.cpp.o
  "__ZNKSt3__14__fs10filesystem4path13__parent_pathEv", referenced from:
      __ZN4SHOT6Solver10setProblemENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in Solver.cpp.o
      __ZN4SHOT18ModelingSystemAMPL13createProblemERNSt3__110shared_ptrINS_7ProblemEEERKNS1_12basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in libSHOTModelingInterfaces.a(ModelingSystemAMPL.cpp.o)
      __ZN4SHOT18ModelingSystemOSiL13createProblemERNSt3__110shared_ptrINS_7ProblemEEERKNS1_12basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in libSHOTModelingInterfaces.a(ModelingSystemOSiL.cpp.o)
  "__ZNKSt3__14__fs10filesystem4path16__root_directoryEv", referenced from:
      __ZN4SHOT6Solver19initializeDebugModeEv in Solver.cpp.o
      __ZNSt3__14__fs10filesystemdvERKNS1_4pathES4_ in libSHOTHelper.a(Utilities.cpp.o)
  "__ZNKSt3__14__fs10filesystem4path6__stemEv", referenced from:
      __ZN4SHOT6Solver10setProblemENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in Solver.cpp.o
  "__ZNKSt3__14__fs10filesystem4path9__compareENS_17basic_string_viewIcNS_11char_traitsIcEEEE", referenced from:
      __ZN4SHOT6Solver10setProblemENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in Solver.cpp.o
  "__ZNSt19bad_optional_accessD1Ev", referenced from:
      __ZNSt3__127__throw_bad_optional_accessEv in libSHOTModel.a(Simplifications.cpp.o)
      __ZNSt3__127__throw_bad_optional_accessEv in libSHOTTasks.a(TaskReformulateProblem.cpp.o)
  "__ZNSt3__14__fs10filesystem10__absoluteERKNS1_4pathEPNS_10error_codeE", referenced from:
      __ZN4SHOT6Solver10setProblemENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in Solver.cpp.o
  "__ZNSt3__14__fs10filesystem11__canonicalERKNS1_4pathEPNS_10error_codeE", referenced from:
      __ZN4SHOT6Solver19initializeDebugModeEv in Solver.cpp.o
  "__ZNSt3__14__fs10filesystem11__copy_fileERKNS1_4pathES4_NS1_12copy_optionsEPNS_10error_codeE", referenced from:
      __ZN4SHOT6Solver19initializeDebugModeEv in Solver.cpp.o
  "__ZNSt3__14__fs10filesystem14__current_pathEPNS_10error_codeE", referenced from:
      __ZN4SHOT6Solver10setProblemENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in Solver.cpp.o
  "__ZNSt3__14__fs10filesystem18__create_directoryERKNS1_4pathEPNS_10error_codeE", referenced from:
      __ZN4SHOT9Utilities24createTemporaryDirectoryENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEES7_ in libSHOTHelper.a(Utilities.cpp.o)
  "__ZNSt3__14__fs10filesystem20__create_directoriesERKNS1_4pathEPNS_10error_codeE", referenced from:
      __ZN4SHOT6Solver19initializeDebugModeEv in Solver.cpp.o
  "__ZNSt3__14__fs10filesystem21__temp_directory_pathEPNS_10error_codeE", referenced from:
      __ZN4SHOT9Utilities24createTemporaryDirectoryENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEES7_ in libSHOTHelper.a(Utilities.cpp.o)
  "__ZNSt3__14__fs10filesystem4path17replace_extensionERKS2_", referenced from:
      __ZN4SHOT18ModelingSystemAMPL13createProblemERNSt3__110shared_ptrINS_7ProblemEEERKNS1_12basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in libSHOTModelingInterfaces.a(ModelingSystemAMPL.cpp.o)
  "__ZNSt3__14__fs10filesystem8__statusERKNS1_4pathEPNS_10error_codeE", referenced from:
      __ZN4SHOT6Solver10setProblemENSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in Solver.cpp.o
      __ZN4SHOT6Solver19initializeDebugModeEv in Solver.cpp.o
      __ZN4SHOT18ModelingSystemAMPL13createProblemERNSt3__110shared_ptrINS_7ProblemEEERKNS1_12basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE in libSHOTModelingInterfaces.a(ModelingSystemAMPL.cpp.o)
  "__ZTINSt3__14__fs10filesystem16filesystem_errorE", referenced from:
      GCC_except_table22 in Solver.cpp.o
  "__ZTISt19bad_optional_access", referenced from:
      __ZNSt3__127__throw_bad_optional_accessEv in libSHOTModel.a(Simplifications.cpp.o)
      __ZNSt3__127__throw_bad_optional_accessEv in libSHOTTasks.a(TaskReformulateProblem.cpp.o)
  "__ZTVSt19bad_optional_access", referenced from:
      __ZNSt3__127__throw_bad_optional_accessEv in libSHOTModel.a(Simplifications.cpp.o)
      __ZNSt3__127__throw_bad_optional_accessEv in libSHOTTasks.a(TaskReformulateProblem.cpp.o)
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
ld: symbol(s) not found for architecture x86_64
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CMakeFiles/SHOTSolver.dir/build.make:148: libSHOTSolver.dylib] Error 1
make[2]: Leaving directory '/workspace/srcdir/SHOT/build'
make[1]: *** [CMakeFiles/Makefile2:451: CMakeFiles/SHOTSolver.dir/all] Error 2
make[1]: Leaving directory '/workspace/srcdir/SHOT/build'
```
I have enough rabbit holes for tonight :disappointed:

**Comment by odow** [comment]
That's windows fixed. Now just to fix Mac...

**Comment by odow** [comment]
@andreaslundell we're building SHOT so that it can be automatically installed by users via JuMP. The interface would be 
```Julia
# This will install SHOT on the user's computer. No need to compile or download the binary separately!
import Pkg; Pkg.add("SHOT_jll")

using JuMP, AmplNLWriter, SHOT_jll

model = Model() do
    return AmplNLWriter.Optimizer(SHOT_jll.amplexe)
end
@variable(model, x, Int)
@NLobjective(model, (x - 0.5)^2)
optimize!(model)
```

Any tips for compiling on Mac? We're having trouble with the more recent filename stuff: https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=12127&view=logs&j=0cdbc6f7-38d4-59bc-31eb-b7a4f3c7d91c&t=94a34fc4-c0e1-5cca-da68-79f57b76f9e5

**Comment by andreaslundell** [comment]
Thanks for taking the time to do this! 

I have never compiled on Mac unfortunately, but I believe @svigerske has since SHOT is working on Mac with GAMS.

**Comment by andreaslundell** [comment]
The compiler warnings at https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=12127&view=logs&j=0cdbc6f7-38d4-59bc-31eb-b7a4f3c7d91c&t=94a34fc4-c0e1-5cca-da68-79f57b76f9e5&l=995.

Have been fixed in the master branch. A new release of SHOT will be released in the coming weeks.

**Comment by svigerske** [comment]
We don't use clang for our build of SHOT on macOS, one reason being the incomplete C++17 support on the old macOS version that we build on.

In the log here, there is an
```
export CXXFLAGS="-mmacosx-version-min=10.14"
```
There are posts like [this one](https://stackoverflow.com/questions/56924853/why-xcode-11-beta-cant-use-c17s-filesystem-header) that indicate that changing this to 10.15 may work, that is, support for `std::filesystem` was added to macOS 10.15, so you cannot use it if you say you want to be compatible with 10.14.

If this is a macOS 10.14 system, then you could try with GCC, whose C++ standard library version isn't tied to the macOS version. That can lead to problems with the Gurobi and CPLEX interfaces, but I guess you won't build these anyway here.

**Comment by giordano** [comment]
We build LLVM with the 10.12 SDK.  I wonder if that precludes all the `std::filesystem` stuff to be available at all, despite using `-mmacosx-version-min=10.14` or `-mmacosx-version-min=10.15` when compiling.

**Comment by giordano** [comment]
Ok, good news: `libc++` comes from the SDK, not our compiler, so we only need to install a newer SDK

**Comment by odow** [comment]
> so we only need to install a newer SDK

Any pointers on how to do this? I guess it's not a simple `apt get`...

**Comment by giordano** [comment]
We do it for Qt, I'm already testing locally whether we need 10.14 or 10.15

**Comment by giordano** [comment]
SDK 10.15 is necessary (and I'm going to push the change), 10.14 didn't work (the `std::filesystem` symbols are still missing there), but now I get:
```
[ 99%] Linking CXX shared library libSHOTSolver.dylib
/usr/bin/cmake -E cmake_link_script CMakeFiles/SHOTSolver.dir/link.txt --verbose=true
/usr/bin/ccache /opt/bin/x86_64-apple-darwin14-libgfortran5-cxx11/x86_64-apple-darwin14-clang++ --sysroot=/opt/x86_64-apple-darwin14/x86_64-apple-darwin14/sys-root -mmacosx-version-min=10.15 -O3 -DNDEBUG -dynamiclib -Wl,-headerpad_max_install_names  -o libSHOTSolver.dylib -install_name @rpath/libSHOTSolver.dylib CMakeFiles/SHOTSolver.dir/src/Report.cpp.o CMakeFiles/SHOTSolver.dir/src/RootsearchMethod/RootsearchMethodBoost.cpp.o CMakeFiles/SHOTSolver.dir/src/Solver.cpp.o  libSHOTHelper.a libSHOTModel.a libSHOTResults.a libSHOTDualStrategy.a libSHOTPrimalStrategy.a libSHOTTasks.a -Wl,-all_load libSHOTModelingInterfaces.a libSHOTSolutionStrategies.a libSHOTTasks.a libSHOTDualStrategy.a -lCbcSolver -lCbc -lpthread -lasl -lCgl -lOsiClp -lClp -lOsi -lCoinUtils libSHOTPrimalStrategy.a libSHOTModel.a libSHOTHelper.a libtinyxml2.a -lipopt 
duplicate symbol '__ZN4SHOT8TaskBase3runEv' in:
    libSHOTHelper.a(TaskBase.cpp.o)
    libSHOTTasks.a(TaskBase.cpp.o)
duplicate symbol '__ZN4SHOT8TaskBase10initializeEv' in:
    libSHOTHelper.a(TaskBase.cpp.o)
    libSHOTTasks.a(TaskBase.cpp.o)
duplicate symbol '__ZN4SHOT8TaskBase8isActiveEv' in:
    libSHOTHelper.a(TaskBase.cpp.o)
    libSHOTTasks.a(TaskBase.cpp.o)
duplicate symbol '__ZN4SHOT8TaskBase10deactivateEv' in:
    libSHOTHelper.a(TaskBase.cpp.o)
    libSHOTTasks.a(TaskBase.cpp.o)
duplicate symbol '__ZN4SHOT8TaskBase8activateEv' in:
    libSHOTHelper.a(TaskBase.cpp.o)
    libSHOTTasks.a(TaskBase.cpp.o)
duplicate symbol '__ZN4SHOT8TaskBase7getTypeEv' in:
    libSHOTHelper.a(TaskBase.cpp.o)
    libSHOTTasks.a(TaskBase.cpp.o)
duplicate symbol '__ZTVN4SHOT8TaskBaseE' in:
    libSHOTHelper.a(TaskBase.cpp.o)
    libSHOTTasks.a(TaskBase.cpp.o)
duplicate symbol '__ZTSN4SHOT8TaskBaseE' in:
    libSHOTHelper.a(TaskBase.cpp.o)
    libSHOTTasks.a(TaskBase.cpp.o)
duplicate symbol '__ZTIN4SHOT8TaskBaseE' in:
    libSHOTHelper.a(TaskBase.cpp.o)
    libSHOTTasks.a(TaskBase.cpp.o)
duplicate symbol '__ZN4SHOT8TaskBaseC2ENSt3__110shared_ptrINS_11EnvironmentEEE' in:
    libSHOTHelper.a(TaskBase.cpp.o)
    libSHOTTasks.a(TaskBase.cpp.o)
duplicate symbol '__ZN4SHOT8TaskBaseC1ENSt3__110shared_ptrINS_11EnvironmentEEE' in:
    libSHOTHelper.a(TaskBase.cpp.o)
    libSHOTTasks.a(TaskBase.cpp.o)
ld: 11 duplicate symbols for architecture x86_64
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CMakeFiles/SHOTSolver.dir/build.make:148: libSHOTSolver.dylib] Error 1
```
This looks to me an upstream bug.

**Comment by odow** [comment]
Great! Yeah I think I found the upstream issue. Testing locally

File: C/Coin-OR/SHOT/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    -DCBC_DIR=${prefix} \
+    -DHAS_IPOPT=on \
+    -DIPOPT_DIR=${prefix} \
+    -DHAS_AMPL=on \
+    -DGENERATE_EXE=on \
+    ..
+
+make -j${nproc}
+make install
+"""
```
Comment (giordano) [inline C/Coin-OR/SHOT/build_tarballs.jl:57]: Ugh, now that we have an additional source the license isn't installed automatically anymore:
```suggestion

install_license ../LICENSE
"""
```
Funnily enough, in `bin/` there are some completely unrelated files:
```console
% ls bin 
AUTHORS  INSTALL  LICENSE  README  SHOT
```

File: C/Coin-OR/SHOT/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+products = [
+    ExecutableProduct("SHOT", :amplexe),
+    LibraryProduct("libSHOTSolver", :libshotsolver),
+]
+
+dependencies = [
+    Dependency("ASL_jll", ASL_version),
+    Dependency("Cbc_jll", Cbc_version),
+    Dependency("Ipopt_jll", Ipopt_version),
```
Comment (giordano) [inline C/Coin-OR/SHOT/build_tarballs.jl:67]: Note that this doesn't define a compat for those versions.  In case you need compat bounds you can do something like
```julia
Dependency("ASL_jll"; compat="$(ASL_version)"),
```
etc...

File: C/Coin-OR/SHOT/bundled/patches/0002-cmake.patch [hunk 5]
```diff
@@ -0,0 +1,13 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
```
Comment (giordano) [inline C/Coin-OR/SHOT/bundled/patches/0002-cmake.patch:1]: Unrelated to this pull request, can this patch be upstreamed?

Comment (odow) [reply C/Coin-OR/SHOT/bundled/patches/0002-cmake.patch:1]: Yes, I think so.

File: C/Coin-OR/SHOT/build_tarballs.jl (hunk 4 above)
Comment (odow) [reply C/Coin-OR/SHOT/build_tarballs.jl:69]: I think this is okay for now. We're going to have a bun fight if/when the new releases of the COIN-OR suite get released, because I'll have to rebuild everything at the same time, and I may as well add compat bounds then.

**Comment by odow** [comment]
All ticks green. There are few warnings:
image
but I think we can safely ignore these?
